### ISSUE TITLE
AWS worker nodes not being auto-assigned an IPv4 address

### ISSUE THREAD

**Issue by JamesBelchamber** [poseidon/typhoon#1173]
Typhoon deployed with the following configuration (barring the `xxxx` placeholders) while reading the documentation [here](https://typhoon.psdn.io/fedora-coreos/aws/):

```
module "kubernetes" {
  source             = "git::https://github.com/poseidon/typhoon//aws/fedora-coreos/kubernetes?ref=v1.24.1"
  cluster_name       = "kubernetes"
  dns_zone           = aws_route53_zone.xxxx.name
  dns_zone_id        = aws_route53_zone.xxxx.id
  ssh_authorized_key = "ssh-rsa xxxx"
  worker_count       = 2
  network_mtu        = 8981
}
```

The output of `kubectl get nodes` lists only the control node. On investigating, the worker nodes are failing to reach the url `https://kubernetes.xxxx.com:6443/` and are experiencing timeouts. This is because the worker nodes have no public IPv4 addresses and there is no NAT Gateway to proxy requests.

Adding a AAAA record that otherwise matches the root A record (alias to the NLB) fixes the issue.

**Comment by dghubble** [comment]
Controller and worker nodes all get public IPv4 addresses ([see](https://github.com/poseidon/typhoon/blob/master/aws/fedora-coreos/kubernetes/network.tf#L55)). Likely your DNS records just hadn't propagated. Check your zone's TTL times https://typhoon.psdn.io/topics/performance/#provision-time

Poseidon actively uses Typhoon clusters before every release. A quad-A record is not neccessary.

**Comment by JamesBelchamber** [comment]
I had the deployment there for 5 hours before troubleshooting, and the worker node could `dig` the endpoint - it was just timing out when connecting. I could also see that the nodes didn't (and still don't) have public IPv4 addresses, because the launch configuration has `associate_public_ip_address` set to false.

**Comment by JamesBelchamber** [comment]
If the worker nodes should have public IP addresses then I guess we could just hardcode `associate_public_ip_address` to true instead. This is possibly a change in the Terraform provider (or AWS themselves) - I'll go have a dig.

**Comment by JamesBelchamber** [comment]
Reasonably sure this has always been like this ([see here](https://github.com/hashicorp/terraform-provider-aws/blame/9f6fa8e10bb0f50d9e2e5d92f9c5be60e663385f/internal/service/autoscaling/launch_configuration.go#L45)). You also explicitly enforce associating Public IP addresses for controller nodes ([here](https://github.com/poseidon/typhoon/blob/1a5949824ca3d9648a1058f94c07366097cd4f68/aws/fedora-coreos/kubernetes/controllers.tf#L37)). I'll raise another PR to enforce associating the public IP address if that's the desired outcome.

My deployment is _very_ minimal so I don't see how I could be getting a different outcome to your testing.

**Comment by dghubble** [comment]
I'm not looking for PRs. There needs to be a reproducible issue and a clear understanding. We have neither. 

It is not neccessary to explicitly `associate_public_ip_address` in launch configurations either. The VPC `map_public_ip_on_launch` should be sufficient. https://docs.aws.amazon.com/vpc/latest/userguide/working-with-subnets.html#subnet-public-ip Please have a look at your VPC, your Terraform AWS provider version, or other factors that may be unique to your AWS account or setup.

**Comment by dghubble** [comment]
Looks like the AWS Terraform provider has a regression introduced 6 days ago in v4.17.0 https://github.com/hashicorp/terraform-provider-aws/issues/25244 I suspected something else was the root cause, since the Typhoon AWS module has been used quite a bit for ages.

I'll have a chat with upstream and see what can be done. For now, you can address this by pinning them to the prior version.

```tf
terraform {
  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "4.16.0"
    }
  }
}
```

**Comment by dghubble** [comment]
Thanks for the report. The AWS Terraform provider fixed the regression in v4.20.0, so the workaround here is not longer needed https://github.com/hashicorp/terraform-provider-aws/issues/25244#issuecomment-1164997502
